### ISSUE TITLE
Installing GTSAM in combination with RoboStack ROS Noetic on Windows 10

### ISSUE THREAD

**Issue by hmaarten** [conda-forge/gtsam-feedstock#25]
### Comment:

I aim to implement a ROS package in Python that lives completely within an Anaconda environment on a Windows computer (we have some other software that must run on Windows). I have been using [RoboStack ROS Noetic](https://robostack.github.io/index.html) to implement the ROS side of things. Additionally, I require GTSAM in one of my ros nodes. After having installed ROS using `mamba install ros-noetic-desktop` I install GTSAM using `mamba install gtsam` which does work, however, it installs version `gtsam-4.2.0-py311h669d342_3` which is not the latest build. When running the following snippet of code (minimal reproducible example)
```
import gtsam
import numpy as np

graph = gtsam.NonlinearFactorGraph()

pose_key = gtsam.symbol('x', 0)
vel_key = gtsam.symbol('v', 0)
bias_key = gtsam.symbol('b', 0)

current_pose = gtsam.Pose3(np.eye(4))
current_vel = np.zeros(3)
current_bias = gtsam.imuBias.ConstantBias()

graph.add(gtsam.PriorFactorPose3(pose_key, current_pose, gtsam.noiseModel.Isotropic.Sigma(6, 0.1)))
graph.add(gtsam.PriorFactorVector(vel_key, current_vel, gtsam.noiseModel.Isotropic.Sigma(3, 0.1)))
graph.add(gtsam.PriorFactorConstantBias(bias_key, current_bias, gtsam.noiseModel.Isotropic.Sigma(6, 0.1)))

initial_estimate = gtsam.Values()
initial_estimate.insert(pose_key, current_pose)
initial_estimate.insert(vel_key, current_vel)
initial_estimate.insert(bias_key, current_bias)

pose_key += 1
vel_key += 1
bias_key += 1

new_pose = gtsam.Pose3(np.eye(4))
graph.add(gtsam.PriorFactorPose3(pose_key, new_pose, gtsam.noiseModel.Diagonal.Sigmas(np.array([0.1, 0.1, 0.1, 0.1, 0.1, 0.1]))))

imu_accum = gtsam.PreintegratedImuMeasurements(gtsam.PreintegrationParams(np.asarray([0, 0, 9.81])))
predicted_nav_state = imu_accum.predict(gtsam.NavState(current_pose, current_vel), current_bias)
imu_factor = gtsam.ImuFactor(pose_key-1, vel_key-1, pose_key, vel_key, bias_key, imu_accum)
graph.add(imu_factor)

initial_estimate.insert(pose_key, predicted_nav_state.pose())
initial_estimate.insert(vel_key, predicted_nav_state.velocity())
initial_estimate.insert(bias_key, gtsam.imuBias.ConstantBias())

bias_factor = gtsam.BetweenFactorConstantBias(bias_key-1, bias_key, gtsam.imuBias.ConstantBias(), gtsam.noiseModel.Diagonal.Sigmas(np.array([0.1, 0.1, 0.1, 0.1, 0.1, 0.1])))
graph.add(bias_factor)

isam = gtsam.ISAM2(gtsam.ISAM2Params())
result = isam.update(graph, initial_estimate)
```
the following error is thrown:
```
RuntimeError: Attempting to retrieve value with key "v1", type stored in Values is class gtsam::GenericValue<class Eigen::Matrix<double,-1,1,0,-1,1> > but requested type was class Eigen::Matrix<double,3,1,0,3,1>
```
This is a known issue and was fixed in [this pull request](https://github.com/borglab/gtsam/pull/1685). Indeed, when installing the latest build of GTSAM (`gtsam-4.2.0-py311hdd3dd35_5`) no error is thrown for the upper code snippet. However, this version is incompatible with `mamba install ros-noetic-desktop`, as versions of `libboost` 1.82 and 1.84 collide.

What could be a workaround to ensure compatibility? Should GTSAM be rebuilt for  `libboost 1.82`, while still incorporating the changes of the above-mentioned pull request? Or is there a combination of versions that is compatible with my requirements and does not throw this error? I have tried installing only a subset of the ROS package, but that didn't work. Also, trying other Python versions yielded no success as ROS Noetic is only built for Python 3.11 as it seems. Any help would be appreciated as I am not well versed with package building using `conda-forge`.

**Comment by traversaro** [comment]
This is unfortunately a recurrent problem with robostack packages. conda-forge has ABI migration occuring regularly (https://conda-forge.org/blog/2019/12/06/cfep09/), but at robostack at the moment we only do periodic rebuilds, that however use the version of dependencies used in conda-forge packages built at the moment we did the rebuild (as you noticed, currently libboost 1.82, see https://github.com/RoboStack/ros-noetic/blob/7f743314408fbaea5701b8f09193e11136b9b899/additional_recipes/ros-distro-mutex/recipe.yaml#L33).

The "proper" solution is to do a rebuild with updated libboost pinning set to 1.84, but that takes time. Other solutions involve separating gtsam and ros code in two separated conda environment, and make them communicate over some kind of non-ROS IPC. Yes another workaround is to compile gtsam from source. 

However, as the boost 1.84 migration is still ongoing (see https://conda-forge.org/status/migration/boost1840) probably at least for gtsam we can do a dual 1.82/1.84 . Doing this for all packages it is not sustainable but for this specific case I do not think it is particularly problematic.

**Comment by traversaro** [comment]
Ok, one step back as I may be a bit confused.

> This is a known issue and was fixed in https://github.com/borglab/gtsam/pull/1685. Indeed, when installing the latest build of GTSAM (gtsam-4.2.0-py311hdd3dd35_5) no error is thrown for the upper code snippet.

If this is issue was fixed in https://github.com/borglab/gtsam/pull/1685, how is this related to gtsam 4.2.0 that was released several months before  https://github.com/borglab/gtsam/pull/1685 was merged? Can you try if you see the same problem with gtsam-4.2.0==*=*_4 ? Thanks!

**Comment by hmaarten** [comment]
Thanks for your response and the explanation. I will check again tomorrow as I don't have my PC but I am quite certain that the behavior is already fixed in `_4` but the same compatibility issues regarding `libboost` arose, i.e. it would only let me install version `_3` alongsid RoboStack ROS. However, the timeline is indeed confusing as  the pull request is not even part of the current build as I understand? Could just upgrading `libboost` to 1.84 have implicity solved the problem (i. e., downgrading would again yield the same error)?

**Comment by traversaro** [comment]
> Could just upgrading `libboost` to 1.84 have implicity solved the problem (i. e., downgrading would again yield the same error)?

If that is the case, then a rebuild against liboost 1.82 will not help.

**Comment by traversaro** [comment]
Just to double check, can you check if you can install suitesparse==7 with the packages you need? If you can't, again the rebuild in https://github.com/conda-forge/gtsam-feedstock/pull/26 will not be useful for you.

**Comment by hmaarten** [comment]
`mamba install -c conda-forge suitesparse=7.7.0` yields
```
Could not solve for environment specs
The following packages are incompatible
├─ gtsam is installable with the potential options
│  ├─ gtsam 4.2.0 would require
│  │  └─ libboost >=1.84.0,<1.85.0a0 , which requires
│  │     └─ boost-cpp 1.84.0* , which can be installed;
│  ├─ gtsam [4.1.1|4.2.0] would require
│  │  └─ suitesparse >=5.4.0,<6.0a0 , which can be installed;
│  ├─ gtsam [4.1.1|4.2.0] would require
│  │  └─ python >=3.10,<3.11.0a0 , which can be installed;
│  ├─ gtsam 4.1.1 would require
│  │  └─ python >=3.7,<3.8.0a0 , which can be installed;
│  ├─ gtsam [4.1.1|4.2.0] would require
│  │  └─ python >=3.8,<3.9.0a0 , which can be installed;
│  ├─ gtsam [4.1.1|4.2.0] would require
│  │  └─ python >=3.9,<3.10.0a0 , which can be installed;
│  └─ gtsam 4.2.0 would require
│     └─ python >=3.12,<3.13.0a0 , which can be installed;
├─ libboost-devel is installable with the potential options
│  ├─ libboost-devel 1.82.0 would require
│  │  └─ libboost 1.82.0 h65993cd_6, which conflicts with any installable versions previously reported;
│  ├─ libboost-devel 1.82.0 would require
│  │  └─ libboost 1.82.0 h65993cd_2, which conflicts with any installable versions previously reported;
│  ├─ libboost-devel 1.82.0 would require
│  │  └─ libboost 1.82.0 h65993cd_3, which conflicts with any installable versions previously reported;
│  ├─ libboost-devel 1.82.0 would require
│  │  └─ libboost 1.82.0 h65993cd_4, which conflicts with any installable versions previously reported;
│  ├─ libboost-devel 1.82.0 would require
│  │  └─ libboost 1.82.0 h65993cd_5, which conflicts with any installable versions previously reported;
│  ├─ libboost-devel [1.83.0|1.84.0], which can be installed;
│  └─ libboost-devel 1.85.0 would require
│     └─ libboost 1.85.0 h9a677ad_0, which conflicts with any installable versions previously reported;
├─ ros-distro-mutex is installable with the potential options
│  ├─ ros-distro-mutex 0.4.0 would require
│  │  └─ boost-cpp 1.78.* , which can be installed;
│  └─ ros-distro-mutex 0.5.0 would require
│     └─ libboost-devel 1.82 , which conflicts with any installable versions previously reported;
├─ ros-noetic-std-msgs is not installable because there are no viable options
│  ├─ ros-noetic-std-msgs 0.5.13 would require
│  │  └─ ros-noetic-message-runtime but there are no viable options
│  │     ├─ ros-noetic-message-runtime 0.4.13 would require
│  │     │  └─ ros-noetic-rostime but there are no viable options
│  │     │     ├─ ros-noetic-rostime 0.7.2 would require
│  │     │     │  └─ boost-cpp but there are no viable options
│  │     │     │     ├─ boost-cpp 1.63.0 would require
│  │     │     │     │  └─ vc 10 , which does not exist (perhaps a missing channel);
│  │     │     │     ├─ boost-cpp [1.63.0|1.64.0|...|1.67.0] would require
│  │     │     │     │  └─ vc 14 , which conflicts with any installable versions previously reported;
│  │     │     │     ├─ boost-cpp [1.63.0|1.64.0|...|1.70.0] would require
│  │     │     │     │  └─ vc [9 |9.* ], which conflicts with any installable versions previously reported;
│  │     │     │     ├─ boost-cpp [1.65.1|1.67.0|...|1.84.0] conflicts with any installable versions previously reported;
│  │     │     │     ├─ boost-cpp 1.82.0 would require
│  │     │     │     │  └─ libboost-devel 1.82.0 h91493d7_6, which cannot be installed (as previously explained);
│  │     │     │     └─ boost-cpp 1.85.0 would require
│  │     │     │        └─ libboost-devel 1.85.0 h91493d7_0, which cannot be installed (as previously explained);
│  │     │     └─ ros-noetic-rostime 0.7.3 would require
│  │     │        └─ ros-distro-mutex 0.5.* , which cannot be installed (as previously explained);
│  │     └─ ros-noetic-message-runtime 0.4.13 would require
│  │        └─ ros-distro-mutex 0.5.* , which cannot be installed (as previously explained);
│  └─ ros-noetic-std-msgs 0.5.13 would require
│     └─ ros-distro-mutex 0.5.* , which cannot be installed (as previously explained);
└─ suitesparse 7.7.0**  is not installable because it conflicts with any installable versions previously reported.

```
so it seems to be incompatible. `suitsparse==5` is the latest version that works. 

So what are my options from here? I guess #26 will not resolve my problems as you mentioned. I assume I would need to build the version [borglab/gtsam/#1685](https://github.com/borglab/gtsam/pull/1685) from the `gtsam` development branch in order to fix my issues still using `libboost 1.82`. How feasible would that be?

**Comment by hmaarten** [comment]
`mamba install -c conda-forge suitesparse=7.7.0` is installable alongside Robostack ROS. So this does not seem to be an issue. Even though #26 will probably not resolve my problems as you mentioned, it might still be worth the try. Otherwise, my remaining option would be to build the version [borglab/gtsam/#1685](https://github.com/borglab/gtsam/pull/1685) from the `gtsam` development branch with `libboost 1.82`. Would that be feasible?

**Comment by traversaro** [comment]
> Would that be feasible?

I do not see any specific problem that could emerge. There are a bunch of patches in https://github.com/conda-forge/gtsam-feedstock/tree/main/recipe, but hopefully they are superseded in master thanks to the PR https://github.com/borglab/gtsam/pull/1685, so a good attempt would just to try to compile the master version as it is is. You can get inspiration from the https://github.com/conda-forge/gtsam-feedstock/blob/main/recipe/bld.bat script, just make sure that you use `%CONDA_PREFIX%` instead of `%PREFIX%` if you build outside of conda-build.  

> Even though https://github.com/conda-forge/gtsam-feedstock/pull/26 will probably not resolve my problems as you mentioned, it might still be worth the try. 

To be honest, I am not sure about that. Not sure if I am missing anything, but issue does not seem to be related to boost.

**Comment by hmaarten** [comment]
Okay, so #26 should be closed then.
Regarding building the GTSAM develop branch, would you suggest following the [python wrapper building instructions](https://github.com/borglab/gtsam/tree/develop/python) (so do not work with conda-build) as I aim to use it only locally? I am not experienced in building packages, so if you have any other pointers on how to start, let me know.
Thanks

**Comment by traversaro** [comment]
Those instructions seems fine, the only thing that is important is not to install any dependency with pip, but rather use conda to install the dependencies listed in https://github.com/conda-forge/gtsam-feedstock/blob/main/recipe/meta.yaml#L30-L53 . Note that instead of the compilers macros, you can just use the `compilers` package, and you can ignore all the lines that end with `# [build_platform != target_platform]`.

**Comment by traversaro** [comment]
I just tested your script, and it is failing with the reported error with the following environment:

~~~
(gtsam) D:\src\gtsamws>conda list
# packages in environment at D:\miniforge3\envs\gtsam:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.8                hcfcfb64_5    conda-forge
ca-certificates           2024.2.2             h56e8100_0    conda-forge
geographiclib-cpp         2.3                  h91493d7_1    conda-forge
gtsam                     4.2.0           py312h8ae2717_5    conda-forge
intel-openmp              2024.1.0           h57928b3_965    conda-forge
libblas                   3.9.0              22_win64_mkl    conda-forge
libboost                  1.84.0               h9a677ad_3    conda-forge
libcblas                  3.9.0              22_win64_mkl    conda-forge
libexpat                  2.6.2                h63175ca_0    conda-forge
libffi                    3.4.2                h8ffe710_5    conda-forge
libflang                  5.0.0           h6538335_20180525    conda-forge
libhwloc                  2.10.0          default_h2fffb23_1000    conda-forge
libiconv                  1.17                 hcfcfb64_2    conda-forge
liblapack                 3.9.0              22_win64_mkl    conda-forge
libsqlite                 3.45.3               hcfcfb64_0    conda-forge
libxml2                   2.12.7               h283a6d9_0    conda-forge
libzlib                   1.2.13               hcfcfb64_5    conda-forge
llvm-meta                 5.0.0                         0    conda-forge
metis                     5.1.0             h63175ca_1007    conda-forge
mkl                       2024.1.0           h66d3029_692    conda-forge
numpy                     1.26.4          py312h8753938_0    conda-forge
openmp                    5.0.0                    vc14_1    conda-forge
openssl                   3.3.0                hcfcfb64_0    conda-forge
pip                       24.0               pyhd8ed1ab_0    conda-forge
pthreads-win32            2.9.1                hfa6e2cd_3    conda-forge
pyparsing                 3.1.2              pyhd8ed1ab_0    conda-forge
python                    3.12.3          h2628c8c_0_cpython    conda-forge
python_abi                3.12                    4_cp312    conda-forge
setuptools                69.5.1             pyhd8ed1ab_0    conda-forge
suitesparse               7.7.0                h7e725d4_1    conda-forge
tbb                       2021.12.0            h91493d7_0    conda-forge
tk                        8.6.13               h5226925_1    conda-forge
tzdata                    2024a                h0c530f3_0    conda-forge
ucrt                      10.0.22621.0         h57928b3_0    conda-forge
vc                        14.3                hcf57466_18    conda-forge
vc14_runtime              14.38.33130         h82b7239_18    conda-forge
vs2015_runtime            14.38.33130         hcb4865c_18    conda-forge
wheel                     0.43.0             pyhd8ed1ab_1    conda-forge
xz                        5.2.6                h8d14728_0    conda-forge
zstd                      1.5.6                h0ea2cb4_0    conda-forge
~~~

Also installing python==3.11.* it still fails. @hmaarten can you please report the exact environment in which the test is working? Thanks!

**Comment by hmaarten** [comment]
Hmm, I just realized that it indeed only works with build `_4`, so the overall dependencies are
```
(gtsam_test) C:\>mamba list
# packages in environment at C:\miniconda3\envs\gtsam_test:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.8                hcfcfb64_5    conda-forge
ca-certificates           2024.2.2             h56e8100_0    conda-forge
geographiclib-cpp         2.3                  h91493d7_1    conda-forge
gtsam                     4.2.0           py311h0631a01_4    conda-forge
intel-openmp              2024.1.0           h57928b3_965    conda-forge
libblas                   3.9.0              22_win64_mkl    conda-forge
libboost                  1.84.0               h9a677ad_3    conda-forge
libcblas                  3.9.0              22_win64_mkl    conda-forge
libexpat                  2.6.2                h63175ca_0    conda-forge
libffi                    3.4.2                h8ffe710_5    conda-forge
libhwloc                  2.10.0          default_h2fffb23_1000    conda-forge
libiconv                  1.17                 hcfcfb64_2    conda-forge
liblapack                 3.9.0              22_win64_mkl    conda-forge
liblapacke                3.9.0              22_win64_mkl    conda-forge
libsqlite                 3.45.3               hcfcfb64_0    conda-forge
libxml2                   2.12.7               h283a6d9_0    conda-forge
libzlib                   1.2.13               hcfcfb64_5    conda-forge
metis                     5.1.0             h63175ca_1007    conda-forge
mkl                       2024.1.0           h66d3029_692    conda-forge
numpy                     1.26.4          py311h0b4df5a_0    conda-forge
openssl                   3.3.0                hcfcfb64_0    conda-forge
pip                       24.0               pyhd8ed1ab_0    conda-forge
pthreads-win32            2.9.1                hfa6e2cd_3    conda-forge
pyparsing                 3.1.2              pyhd8ed1ab_0    conda-forge
python                    3.11.9          h631f459_0_cpython    conda-forge
python_abi                3.11                    4_cp311    conda-forge
setuptools                69.5.1             pyhd8ed1ab_0    conda-forge
suitesparse               5.4.0                h5d0cbe0_1    conda-forge
tbb                       2021.12.0            h91493d7_0    conda-forge
tk                        8.6.13               h5226925_1    conda-forge
tzdata                    2024a                h0c530f3_0    conda-forge
ucrt                      10.0.22621.0         h57928b3_0    conda-forge
vc                        14.3                hcf57466_18    conda-forge
vc14_runtime              14.38.33130         h82b7239_18    conda-forge
vs2015_runtime            14.38.33130         hcb4865c_18    conda-forge
wheel                     0.43.0             pyhd8ed1ab_1    conda-forge
xz                        5.2.6                h8d14728_0    conda-forge
zstd                      1.5.6                h0ea2cb4_0    conda-forge
```

The main difference seems to be the version of `suitesparse` (5.4.0 vs 7.7.0). With this environment, my script works fine...
No idea how to interpret this behavior.

**Comment by hmaarten** [comment]
On another note, I tried to build gtsam from source using changes in [borglab/gtsam/#1685](https://github.com/borglab/gtsam/pull/1685). I tried two approaches

1. Managing the build process from within an Anaconda environment. For this I
-  Installed dependencies from https://github.com/conda-forge/gtsam-feedstock/blob/main/recipe/meta.yaml#L30-L53 using `mamba install` (wiht boost 1.82)
- Tried to apply patches using `git apply ...`, most of them resulted in `patch does not apply` though (I think as there are changes already in the branch)
- Used https://github.com/conda-forge/gtsam-feedstock/blob/main/recipe/bld.bat as a starting point for `cmake`

However, here already `cmake` failed as it could not find `boost` (it did identify version 1.82, but said a bunch of subcomponents were missing, even though they are all present in the environments Library/include and Library/lib folder, respectively. So not sure what to do with that. Here I did not understand exactly what to do with `%LIBRARY_PREFIX%` variable, I just set it to the Library folder of my anaconda environment using `set` at the beginning of the .bat script. @traversaro you mention

> just make sure that you use %CONDA_PREFIX% instead of %PREFIX% if you build outside of conda-build.

Could you elaborate on that?

2. Managing the build process outside of anaconda. For this I
- Installed boost 1.82 using precompiled binaries from the official website and the other dependencies using pip
- Tried to apply patches using `git apply ...`, most of them resulted in `patch does not apply` though (I think as there are changes already in the branch)
- Used [python wrapper building instructions](https://github.com/borglab/gtsam/tree/develop/python) for `cmake`, yielding an .sln file
- Used MSVS 2019 to build the `python-test` package
- Installed it using `pip installl .` in the build/python folder

This approach almost works, however when I try to import gtsam it throws
```
>>> import gtsam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\hmaarten\Documents\gtsam_build\gtsam_with_patches\gtsam\build\python\gtsam\__init__.py", line 9, in <module>
    from gtsam import gtsam, utils
ImportError: cannot import name 'gtsam' from partially initialized module 'gtsam' (most likely due to a circular import) (C:\Users\hmaarten\Documents\gtsam_build\gtsam_with_patches\gtsam\build\python\gtsam\__init__.py)
```
I found a similar issue here #19, and the workaround is included in the https://github.com/conda-forge/gtsam-feedstock/blob/main/recipe/bld.bat file with 
`python\gtsam\gtsam.*.pyd "%SP_DIR%\gtsam\"`
i.e., copying some `.pyd` files. However, when I look in my output build folder, I only find a `.pyd` file in `gtsam\build\python\gtsam\Release` (gtsam_py.pyd), so it does not really match the structure of that command. Maybe because I used MSVS 2019 instead of `ninja` in the 2nd approach.

So currently I am a bit stuck with both approaches. Any additional inputs would be appreciated.

**Comment by traversaro** [comment]
I reproduced the problem in CI in https://github.com/conda-forge/gtsam-feedstock/pull/26 . Let's just backport the patch and if that works and live happy.

**Comment by traversaro** [comment]
> here already `cmake` failed as it could not find `boost` (it did identify version 1.82, but said a bunch of subcomponents were missing, even though they are all present in the environments Library/include and Library/lib folder, respectively.

Please, always include the exact error that you are experiencing, not a recap. The exact errors are important to provide you feedback.

**Comment by traversaro** [comment]
> > just make sure that you use %CONDA_PREFIX% instead of %PREFIX% if you build outside of conda-build.
> 
> Could you elaborate on that?

conda-build works with two environments, a build and host environment. If you do a local build, you just have an environment, pointed by `CONDA_PREFIX`. `LIBRARY_PREFIX`,`PREFIX` and similar are all conda-build env variables, see https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html .

**Comment by traversaro** [comment]
> Maybe because I used MSVS 2019 instead of `ninja` in the 2nd approach.

Yes, that is the reason. VS is a so-called multiple config generator, while `ninja` is a single config, so all the outpus of VS generator have an additional folder to identify the config. Just change the copy command to handle the difference folder structure, or use ninja .

**Comment by traversaro** [comment]
Can you try to set `CMAKE_PREFIX_PATH` cmake variable to `C:\Users\hmaarten\miniforge3\envs\gtsam_build\Library` ?

> I tried copying the `.pyd` file from `gtsam\build\python\gtsam\Release` to `miniconda3\envs\ros_noetic_minimal\Lib\site-packages\gtsam\Release`, this did not yield any change, however.

Any change w.r.t. to what?

**Comment by hmaarten** [comment]
Changed it now to 
```
set CONDA_PREFIX=C:\Users\hmaarten\miniforge3\envs\gtsam_build
set CONDA_PY=C:\Users\hmaarten\miniforge3\envs\gtsam_build\python.exe

set CMAKE_PREFIX_PATH=C:\Users\hmaarten\miniforge3\envs\gtsam_build\Library
set SRC_DIR=C:\Users\hmaarten\Documents\gtsam_build\gtsam 
set LIBRARY_PREFIX=%CONDA_PREFIX%\Library
set PYTHON=%CONDA_PY%
```
which did not change anything of the cmake output (still boost error)

> Any change w.r.t. to what?

As in the circular import issue 
```
>>> import gtsam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\hmaarten\Documents\gtsam_build\gtsam_with_patches\gtsam\build\python\gtsam\__init__.py", line 9, in <module>
    from gtsam import gtsam, utils
ImportError: cannot import name 'gtsam' from partially initialized module 'gtsam' (most likely due to a circular import) (C:\Users\hmaarten\Documents\gtsam_build\gtsam_with_patches\gtsam\build\python\gtsam\__init__.py)
```
remains. Using ninja instead for the build outside of anaconda, i.e., changing

`cmake .. -DGTSAM_BUILD_PYTHON=1 -DGTSAM_PYTHON_VERSION=3.11`

to 

`cmake .. -GNinja -DGTSAM_BUILD_PYTHON=1 -DGTSAM_PYTHON_VERSION=3.11`

yields

```
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The CXX compiler identification is GNU 13.1.0
-- The C compiler identification is GNU 13.1.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Strawberry/c/bin/c++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Strawberry/c/bin/gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- GTSAM is a shared library due to GTSAM_FORCE_SHARED_LIB
-- GTSAM_POSE3_EXPMAP=ON, enabling GTSAM_ROT3_EXPMAP as well
-- Performing Test COMPILER_HAS_WSUGGEST_OVERRIDE
-- Performing Test COMPILER_HAS_WSUGGEST_OVERRIDE - Success
-- Performing Test COMPILER_HAS_WMISSING_OVERRIDE
-- Performing Test COMPILER_HAS_WMISSING_OVERRIDE - Failed
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Could NOT find Boost (missing: Boost_INCLUDE_DIR serialization system filesystem thread program_options date_time timer chrono regex) (Required is at least version "1.65")
CMake Error at cmake/HandleBoost.cmake:33 (message):
  Missing required Boost components >= v1.65, please install/upgrade Boost or
  configure your search paths.
Call Stack (most recent call first):
  CMakeLists.txt:66 (include)

-- Configuring incomplete, errors occurred!
```

so indicating again a `boost` not found error. (without -GNinja flag boost was found and `cmake` is successful)

Thanks for your continued support, I much appreciate it.

**Comment by hmaarten** [comment]
Problem is fixed for me with the latest build, thanks a lot!

**Comment by hmaarten** [comment]
Problem is fixed for me with the latest build #26 , thanks a lot!
